### PR TITLE
Implement convex regions

### PR DESCRIPTION
**Pull request by sethrj**
This adds a few of the fundamental convex regions needed to build Geant4 (see #1078). Like Geant4 and unlike SCALE, these regions are all z-aligned and centered on the origin: it's expected that SCALE shapes will combine these primitives with a Transform (translation and/or permutation).
- Box
- Prism (extruded regular polygon)

DEFERRED to a follow-on commit:
- Cone
- Cylinder
- Ellipsoid
- Sphere

Each of these regions implements:
- Surface construction (automatically named and transformed using the convex surface builder #1113)
- Exterior bounding box construction
- Maximal-volume interior bounding box construction

Not implemented for now:
- Volume calculation
- Other convex regions (elliptical cylinder, half-hyperbola, elliptical cone, trapezoid, tetrahedron, parallelepiped, ...)

### CODE REVIEW

**Comment by sethrj** [comment]
@elliottbiondo This is ready for review. I can potentially break a few changes off into another PR if this one is too big... but the majority of this PR is a single chunk (https://github.com/celeritas-project/celeritas/pull/1119/files#diff-2798c1f414953f0e9b1e93f42727bcec5c6a8ce0da38a65d39099838fd923903 and its test).

**Comment by elliottbiondo** [comment]
Would you mind splitting it up? For reasons I can fully explain, two 700 line merge requests are somehow easier to review than a single 1400 line one.

**Comment by sethrj** [comment]
No need for an explanation; I already know 😄  most of the delta is actually a 500-line unit test 😉

**Comment by elliottbiondo** [comment]
Oops I meant to say "can't fully explain". Will review this soon.

File: src/orange/orangeinp/ConvexRegion.hh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+class ConvexSurfaceBuilder;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Interface class for building non-reentrant spatial regions.
+ *
+ * This is a building block for constructing more complex objects out of
+ * smaller spatial regions. A \c shape object will have a single convex region,
+ * and a \c solid object region may have multiple adjacent convex regions.
+ *
```
Comment (elliottbiondo) [inline src/orange/orangeinp/ConvexRegion.hh:28]: Mention that convex regions are expected to be z-aligned and centered on the origin

File: src/orange/orangeinp/ConvexRegion.hh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * This is the base component of a G4Polyhedra (PGON). The default rotation is
+ * to put a y-aligned plane on the bottom of the shape, so looking at an x-y
+ * slice given an apothem \em a, every shape has a surface at \f$ y = -a \f$:
+ * - n=3 is a triangle with a flat bottom, point up
+ * - n=4 is a square with axis-aligned sides
+ * - n=6 is a flat-top hexagon
+ *
+ * The "orientation" parameter is a scaled counterclockwise rotation on
+ * \f$[0, 1)\f$, where zero preserves the orientation described above, and
+ * unity replicates the original shape but with faces indexed by one. The usual
```
Comment (elliottbiondo) [inline src/orange/orangeinp/ConvexRegion.hh]: Indexed or decremented (as you say in the .cc)?

File: src/orange/orangeinp/ConvexRegion.hh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * This is a building block for constructing more complex objects out of
+ * smaller spatial regions. A \c shape object will have a single convex region,
+ * and a \c solid object region may have multiple adjacent convex regions.
+ *
+ * When implementing this class, prefer to build simpler surfaces (planes)
+ * before complex ones (cones) in case we implement short-circuiting logic,
+ * since expressions are currently sorted.
+ *
+ * \todo Should this be named an "interface"?
```
Comment (elliottbiondo) [inline src/orange/orangeinp/ConvexRegion.hh]: Remove todo

File: src/orange/orangeinp/detail/InfixStringBuilder.hh [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Transform a CSG node into a string expression.
```
Comment (elliottbiondo) [inline src/orange/orangeinp/detail/InfixStringBuilder.hh]: Can you either document here, or reference, the format of the string, i.e. what each symbol denotes, possible with a full example?

File: src/orange/orangeinp/detail/InfixStringBuilder.hh [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    static_assert(to_sense(true) == Sense::outside);
+    *os_ << (negated_ ? '-' : '+') << s.id.unchecked_get();
+    negated_ = false;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Push an aliased node.
+ *
+ * TODO: aliased node shouldn't be reachable if we're fully simplified.
```
Comment (elliottbiondo) [inline src/orange/orangeinp/detail/InfixStringBuilder.hh]: Would it be useful to throw some developer-level warning in this case?

File: src/orange/orangeinp/ConvexSurfaceBuilder.cc [hunk 6]
```diff
@@ -129,11 +130,53 @@ void ConvexSurfaceBuilder::insert_transformed(std::string&& extension,
     state_->nodes.push_back(node_id);
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Shrink the exterior local bounding box.
```
Comment (elliottbiondo) [inline src/orange/orangeinp/ConvexSurfaceBuilder.cc]: local and global here and below

**Review by elliottbiondo — APPROVED** [review]

File: src/orange/orangeinp/ConvexRegion.hh (hunk 3 above)
Comment (sethrj) [reply src/orange/orangeinp/ConvexRegion.hh]: Do you think it should be? It is a pure abstract base class which (generally?) we name `Interface`.

File: src/orange/orangeinp/detail/InfixStringBuilder.hh (hunk 5 above)
Comment (sethrj) [reply src/orange/orangeinp/detail/InfixStringBuilder.hh]: Updated the docs: for testing, not all nodes will be fully simplified.

File: src/orange/orangeinp/detail/InfixStringBuilder.hh (hunk 4 above)
Comment (sethrj) [reply src/orange/orangeinp/detail/InfixStringBuilder.hh]: Added documentation.

File: src/orange/orangeinp/ConvexRegion.hh (hunk 3 above)
Comment (sethrj) [reply src/orange/orangeinp/ConvexRegion.hh]: Derp I already changed it.

**Pull request merged**